File: bee-content/video-editor/src/bee_video_editor/services/matcher.py
"""Media matching service — auto-assign media files to storyboard segments.

Scores available media files against segment descriptions and assigns
best matches. Works without an LLM using keyword overlap + fuzzy matching.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

from bee_video_editor.formats.parser import ParsedStoryboard, ParsedSegment


@dataclass
class Assignment:
    segment_id: str
    layer: str
    layer_index: int
    file_path: Path
    confidence: float  # 0.0-1.0
    reason: str


@dataclass
class AssignmentPlan:
    assignments: list[Assignment] = field(default_factory=list)
    unmatched: list[str] = field(default_factory=list)
    conflicts: list[str] = field(default_factory=list)


@dataclass
class _MediaFile:
    path: Path
    name: str
    tokens: set[str]
    media_type: str  # "video" or "photo"


_STOP_WORDS = {
    "a", "an", "the", "of", "in", "on", "at", "to", "for", "is", "it",
    "and", "or", "but", "with", "from", "by", "as", "into", "that", "this",
    "be", "are", "was", "were", "been", "being", "have", "has", "had",
    "do", "does", "did", "will", "would", "could", "should", "may", "might",
    "no", "not", "very", "just", "all", "also", "then", "than", "so",
    "up", "out", "off", "over", "under", "about", "between",
}

_MEDIA_EXTENSIONS = {".mp4", ".mkv", ".avi", ".mov", ".webm", ".jpg", ".jpeg", ".png", ".webp"}
_VIDEO_EXTENSIONS = {".mp4", ".mkv", ".avi", ".mov", ".webm"}


def _tokenize(text: str) -> set[str]:
    words = re.split(r'[^a-zA-Z0-9]+', text.lower())
    return {w for w in words if w and len(w) > 2 and w not in _STOP_WORDS}


def _build_inventory(media_dirs: list[Path]) -> list[_MediaFile]:
    inventory: list[_MediaFile] = []
    for media_dir in media_dirs:
        if not media_dir.exists():
            continue
        for f in sorted(media_dir.rglob("*")):
            if f.is_file() and f.suffix.lower() in _MEDIA_EXTENSIONS:
                inventory.append(_MediaFile(
                    path=f,
                    name=f.stem.lower(),
                    tokens=_tokenize(f.stem),
                    media_type="video" if f.suffix.lower() in _VIDEO_EXTENSIONS else "photo",
                ))
    return inventory


def _segment_description(seg: ParsedSegment) -> str:
    """Build a description string from segment data for keyword matching."""
    parts = [seg.title]
    for v in seg.config.visual:
        if v.query:
            parts.append(v.query)
        if v.src:
            parts.append(Path(v.src).stem)
        if v.text:
            parts.append(v.text)
    if seg.narration:
        parts.append(seg.narration[:200])  # first 200 chars of narration
    return " ".join(parts)


def _try_src_match(seg: ParsedSegment, inventory: list[_MediaFile]) -> Assignment | None:
    """Try to match a segment's visual src to an inventory file."""
    for i, v in enumerate(seg.config.visual):
        if not v.src:
            continue
        ref_name = Path(v.src).stem.lower()
        for mf in inventory:
            if ref_name in mf.name or mf.name in ref_name:
                return Assignment(
                    segment_id=seg.id, layer="visual", layer_index=i,
                    file_path=mf.path, confidence=0.95,
                    reason=f"src match: {v.src}",
                )
    return None


def _find_best_keyword_match(
    segment_id: str, layer: str, layer_index: int,
    description: str, inventory: list[_MediaFile],
    used_files: dict[Path, int],
) -> Assignment | None:
    desc_tokens = _tokenize(description)
    if not desc_tokens:
        return None

    best_score = 0.0
    best_file: _MediaFile | None = None

    for mf in inventory:
        if not mf.tokens:
            continue
        overlap = desc_tokens & mf.tokens
        if not overlap:
            continue
        score = len(overlap) / min(len(desc_tokens), len(mf.tokens))
        uses = used_files.get(mf.path, 0)
        if uses > 0:
            score *= max(0.3, 1.0 - (uses * 0.2))
        if score > best_score:
            best_score = score
            best_file = mf

    if best_file and best_score > 0:
        return Assignment(
            segment_id=segment_id, layer=layer, layer_index=layer_index,
            file_path=best_file.path, confidence=min(best_score, 1.0),
            reason=f"keyword match ({best_score:.2f}): {', '.join(desc_tokens & best_file.tokens)}",
        )
    return None


def auto_assign_media(
    parsed: ParsedStoryboard,
    media_dirs: list[Path],
    *,
    min_confidence: float = 0.1,
) -> AssignmentPlan:
    """Auto-assign media files to storyboard segments."""
    plan = AssignmentPlan()
    inventory = _build_inventory(media_dirs)
    if not inventory:
        plan.unmatched = [seg.id for seg in parsed.segments]
        return plan

    used_files: dict[Path, int] = {}

    for seg in parsed.segments:
        matched = False

        # Strategy 1: src match (visual entries already have src paths)
        assignment = _try_src_match(seg, inventory)
        if assignment and assignment.confidence >= min_confidence:
            plan.assignments.append(assignment)
            used_files[assignment.file_path] = used_files.get(assignment.file_path, 0) + 1
            matched = True

        # Strategy 2: keyword match using segment description
        if not matched:
            description = _segment_description(seg)
            best = _find_best_keyword_match(
                seg.id, "visual", 0, description, inventory, used_files,
            )
            if best and best.confidence >= min_confidence:
                plan.assignments.append(best)
                used_files[best.file_path] = used_files.get(best.file_path, 0) + 1
                matched = True

        if not matched:
            plan.unmatched.append(seg.id)

    for path, count in used_files.items():
        if count > 3:
            plan.conflicts.append(f"{path.name} used {count} times")

    return plan

File: bee-content/video-editor/web/server/services/matcher.ts
import fs from 'node:fs';
import path from 'node:path';
import type { BeeProject } from '../../shared/types.js';

// --- Tokenization ---

const STOPWORDS = new Set([
  'a', 'an', 'the', 'of', 'in', 'on', 'at', 'to', 'for', 'is', 'it',
  'and', 'or', 'but', 'not', 'no', 'by', 'with', 'from', 'as', 'be',
  'was', 'were', 'are', 'been', 'has', 'had', 'have', 'do', 'does',
  'did', 'will', 'can', 'may', 'this', 'that', 'into', 'over', 'then',
  'than', 'its', 'his', 'her', 'our', 'all', 'also', 'just', 'about',
]);

const MEDIA_EXTS = new Set([
  'mp4', 'mkv', 'avi', 'mov', 'webm', 'jpg', 'jpeg', 'png', 'webp',
  'mp3', 'wav', 'aac', 'm4a', 'ogg', 'flac',
]);

export function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter(w => w.length >= 3 && !STOPWORDS.has(w));
}

// --- File scanning ---

function scanMediaFiles(dirs: string[]): Array<{ absPath: string; relPath: string; tokens: string[] }> {
  const files: Array<{ absPath: string; relPath: string; tokens: string[] }> = [];

  function walk(dir: string, baseDir: string) {
    let entries: string[];
    try {
      entries = fs.readdirSync(dir);
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.startsWith('.')) continue;
      const absPath = path.join(dir, entry);
      try {
        const stat = fs.statSync(absPath);
        if (stat.isDirectory()) {
          walk(absPath, baseDir);
        } else if (stat.isFile()) {
          const ext = path.extname(entry).slice(1).toLowerCase();
          if (!MEDIA_EXTS.has(ext)) continue;
          const relPath = path.relative(baseDir, absPath);
          const nameNoExt = path.basename(entry, path.extname(entry));
          files.push({ absPath, relPath, tokens: tokenize(nameNoExt) });
        }
      } catch {
        continue;
      }
    }
  }

  for (const dir of dirs) {
    walk(dir, path.dirname(dir));
  }

  return files;
}

// --- Scoring ---

function scoreMatch(descTokens: string[], fileTokens: string[]): number {
  if (descTokens.length === 0 || fileTokens.length === 0) return 0;
  const fileSet = new Set(fileTokens);
  let overlap = 0;
  for (const t of descTokens) {
    if (fileSet.has(t)) overlap++;
  }
  return overlap / Math.min(descTokens.length, fileTokens.length);
}

// --- Auto-assign ---

export interface AssignmentResult {
  assigned: number;
  unmatched: number;
  conflicts: string[];
}

export function autoAssignMedia(
  project: BeeProject,
  projectDir: string,
  minConfidence: number = 0.1,
): AssignmentResult {
  // Scan media directories
  const mediaDirs = [
    'footage', 'stock', 'photos', 'graphics', 'narration',
    'maps', 'music', 'generated',
  ].map(d => path.join(projectDir, d));

  const inventory = scanMediaFiles(mediaDirs);
  const useCounts = new Map<string, number>();
  let assigned = 0;
  let unmatched = 0;
  const conflicts: string[] = [];

  for (const seg of project.segments) {
    for (let i = 0; i < seg.visual.length; i++) {
      const entry = seg.visual[i];
      if (entry.src) continue; // already assigned

      // Build description from segment title, section, query, type
      const descText = [seg.title, seg.section, entry.query || '', entry.type || ''].join(' ');
      const descTokens = tokenize(descText);

      let bestFile: (typeof inventory)[0] | null = null;
      let bestScore = 0;

      for (const file of inventory) {
        let score = scoreMatch(descTokens, file.tokens);
        if (score < minConfidence) continue;

        // Reuse penalty
        const uses = useCounts.get(file.relPath) ?? 0;
        score *= Math.max(0.3, 1.0 - uses * 0.2);

        if (score > bestScore) {
          bestScore = score;
          bestFile = file;
        }
      }

      if (bestFile) {
        entry.src = bestFile.relPath;
        const uses = (useCounts.get(bestFile.relPath) ?? 0) + 1;
        useCounts.set(bestFile.relPath, uses);
        if (uses > 3 && !conflicts.includes(bestFile.relPath)) {
          conflicts.push(bestFile.relPath);
        }
        assigned++;
      } else {
        unmatched++;
      }
    }
  }

  return { assigned, unmatched, conflicts };
}
